File: go/primes.go
import (
  "math"
)

func factors(n int) map[int]int {
  m := n
  s := int(math.Sqrt(float64(n)))
  var a []int
  b := [3]int{2, 3, 5}
  c := [8]int{4, 2, 4, 2, 4, 6, 2, 6}
  d := make(map[int]int)
  p := 7
  i := 0

  if n < 2 {
    d[n] = 1
    return d
  }

  for _,p := range b {
    for m%p == 0 {
      a = append(a,p)
      m /= p
    }
  }

  for m != 1 || p <= s {
    for m%p == 0 {
      a = append(a,p)
      m /= p
    }
    p += c[i]
    i = (i+1)%8
  }
  for _,q := range a {
    d[q]++
  }
 return d
}


func IsPrime(n int) bool {
  if n<2 {return false}
  if n==2 || n==3 || n==5 || n==7 {return true}
  if (n%2)==0||(n%3)==0||(n%5)==0||(n%7)==0 {return false}
  c := [8]int{4, 2, 4, 2, 4, 6, 2, 6}
  p := 7
  i := 0
  for p*p <= n {
    if (n%p)==0 {return false}
    p += c[i]
    i = (i+1)%8
  }
 return true
}
func NextPrime (n int) int {
  if n<2 {return 2}
  if n==2 {return 3}
  if (n%2)==0 {
    n++
  } else {n+=2}
  for true {
    if IsPrime(n) {
      return n
    } else {n+=2}
  }
  return -1
}


import "math/big"

func isPrime(n int) bool { return big.NewInt(int64(n)).ProbablyPrime(0) }



package kata



import (
  "math"
)

var primes = []int{2,3}
var h = make(map[int]bool)

func makePrimes() {
  if len(primes) > 2 { return }
  n := 30000
  var c int
  if (n % 6) > 1 { c = 1 }
  n = []int{n, n - 1, n + 4, n + 3, n + 2, n + 1}[n%6]
  ls := make([]bool, n/3)
  
  for i := 1 ; i < n / 3 ; i++ {
    ls[i] = true
  }
  sq := int(math.Sqrt(float64(n)))
  for i := 0 ; i <= sq / 3 ; i++ {
    if ls[i] {
      j := (3*i+1)|1
      h := j * j
      for e := h / 3 ; e < len(ls) ; e += 2 * j {
        ls[e] = false
      }
      for e := (h+4*j-2*j*(i%2))/3 ; e < len(ls) ; e += 2 * j {
        ls[e] = false
      }
    }
  }
  for i := 1 ; i < n/3 - c ; i++ {
    if ls[i] {
      primes = append(primes, (3*i+1)|1)
    }
  }
  for _,p := range primes { h[p] = true }
}

func BigPrimefacDiv(n int) []int {
  makePrimes()
  if n < 0 { n = -n }
  if h[n] { return nil }
  var i, x, y int
  var flag bool
  p := primes[0]
  for p * p <= n {
    p = primes[i]
    i++
    if n % p == 0 {
      y = p
      if ! flag {
        n /= p
        x = n
        flag = true
      }
      for n % p == 0 { n /= p }
    }
  }
  if x == 0 { return nil }
  if n == 1 { return []int{y, x} }
  return []int{n, x}
}

File: haskell/primes.hs
factors :: Integer -> [Integer]
factors n | n < 2 = []
          | even n = 2 : factors (div n 2)
          | null a = [n]
          | otherwise = h : factors (div n h)
  where
    a = filter(\x -> mod n x == 0) [3,5..floor $ sqrt $ fromIntegral n]
    h = head a 
-- https://www.codewars.com/kata/55f0b69fe3ef582c4100008a/solutions/haskell

isPrime :: Integer -> Bool
isPrime x
  | x < 2        = False
  | x == 2       = True
  | mod x 2 == 0 = False
  | otherwise    = all (\p -> mod x p /= 0) [3,5..s]
    where s = ceiling $ sqrt $ fromIntegral x

nextPrime :: Integer -> Integer
nextPrime n = head $ filter isPrime [n+1..]

-- sum by factors
-- https://www.codewars.com/kata/54d496788776e49e6b00052f/solutions/haskell
-- https://www.codewars.com/kata/59ab0ca4243eae9fec000088/solutions/haskell
-- isPrime
-- https://www.codewars.com/kata/5262119038c0985a5b00029f/solutions
-- https://www.codewars.com/kata/5507309481b8bd3b7e001638/solutions/haskell
-- https://www.codewars.com/kata/560b8d7106ede725dd0000e2/solutions/haskell
-- nextPrime
-- https://www.codewars.com/kata/58e230e5e24dde0996000070/solutions/haskell
-- not primes
-- https://www.codewars.com/kata/5a9a70cf5084d74ff90000f7/solutions/haskell
-- https://www.codewars.com/kata/5a9996fa8e503f2b4a002e7a/solutions
-- backward read primes
-- https://www.codewars.com/kata/5539fecef69c483c5a000015/solutions/haskell
-- step in primes
-- https://www.codewars.com/kata/5613d06cee1e7da6d5000055/solutions/haskell
-- sieve
-- https://www.codewars.com/kata/5a908da30025e995880000e3/solutions/haskell
-- https://www.codewars.com/kata/5a9078e24a6b340b340000b8/solutions/haskell
-- https://www.codewars.com/kata/521ef596c106a935c0000519/solutions/haskell
-- totient
-- https://www.codewars.com/kata/53c9157c689f841d16000c03/solutions/haskell
-- kprimes
-- https://www.codewars.com/kata/5726f813c8dcebf5ed000a6b/solutions/haskell
-- factorial decomposition
-- https://www.codewars.com/kata/5a045fee46d843effa000070/solutions/haskell



isPrime :: Integer -> Bool
isPrime x
  | x < 2        = False
  | x == 2       = True
  | mod x 2 == 0 = False
  | otherwise    = all (\p -> mod x p /= 0) [3,5..s]
    where s = ceiling $ sqrt $ fromIntegral x
primeFactors :: Integer -> [(Integer,Integer)]
primeFactors n | n < 2 = []
               | otherwise = b ++ (f' n r 7 0 (cycle [4, 2, 4, 2, 4, 6, 2, 6]))
                where (r,b) = (f n [2,3,5] []) 

f :: Integer -> [Integer] -> [(Integer,Integer)] -> (Integer, [(Integer,Integer)])
f n [] a = (n,a)
f n (p:e) a | mod n p /= 0 = f n e a 
            | otherwise = f r e (a++[(p,x)])
                where (r,x) = red (div n p) p 1 

red :: Integer -> Integer -> Integer -> (Integer,Integer)
red n p i | mod n p == 0 = red (div n p) p (i+1)
          | otherwise = (n,i)

f' :: Integer -> Integer -> Integer -> Int -> [Integer] -> [(Integer,Integer)]
f' n m p i c | p * p > n = if m == n then [(n,1)] else if m > 1 then [(m,1)] else []                                                                                                         
            | m ==1 = []
            | mod m p /= 0 = f' n m (p+(c!!i)) (i+1) c
            | otherwise = [(p,k)] ++ (f' n r (p+(c!!i)) (i+1) c)
                where (r,k) = red (div m p) p 1 

